test/libyul/yulOptimizerTests/fakeStackLimitEvader/connected.yul

```diff
@@ -0,0 +1,62 @@
+{
+	mstore(0x40, memoryguard(0))
+	function g() -> a, b {
+		a := 21
+		let $c := 1
+		b,a,$c := z()
+	}
+	function f() -> x {
+		let $x2
+		$x2 := 42
+		let $x3, $x4 := g()
+		x := mul(add($x2, $x3), h($x4))
+		sstore($x3, $x4)
+	}
+	function h(v) -> a {
+		let x, $z, y := z()
+		a, $z, v := z()
+	}
+	function z() -> a,b,c { let $x := 0 }
+	sstore(0, f())
+	let x, y := g()
+}
+// ----
+// step: fakeStackLimitEvader
+//
+// {
+//     mstore(0x40, memoryguard(0xa0))
+//     function g() -> a, b
+//     {
+//         a := 21
+//         mstore(0x20, 1)
+//         let b_1, a_2, $c_3 := z()
+//         mstore(0x20, $c_3)
+//         a := a_2
+//         b := b_1
+//     }
+//     function f() -> x
+//     {
+//         mstore(0x60, 0)
+//         mstore(0x60, 42)
+//         let $x3_4, $x4_5 := g()
+//         mstore(0x80, $x4_5)
+//         mstore(0x40, $x3_4)
+//         x := mul(add(mload(0x60), mload(0x40)), h(mload(0x80)))
+//         sstore(mload(0x40), mload(0x80))
+//     }
+//     function h(v) -> a_1
+//     {
+//         let x_2_6, $z_7, y_8 := z()
+//         mstore(0x20, $z_7)
+//         let y := y_8
+//         let x_2 := x_2_6
+//         let a_1_9, $z_10, v_11 := z()
+//         mstore(0x20, $z_10)
```

Comment:
In these cases, `z_7` and `z_10` still occupy a stack slot, don't they?

Reply:
Yeah, well, it's a bit complicated :-). But it still works:

Just consider the transformation of ``let x, $z, y := z()`` on its own.

What we do is generate temporary variables instead. That's ``x_2_6, $z_7, y_8`` and first assign to those:

```
let x_2_6, $z_7, y_8 := z()
```

As long as we don't have more than 16 such return variables (which will be handled in subsequent PRs) this will be fine on the stack for now.

Then we move everything to memory that we have to move:

```
mstore(0x20, $z_7)
```

This will mark the stack slot used by ``$z_7`` as available for re-use.

Now we assign to the remaining variables in reversed order:

```
let y := y_8
```
y will reuse the stack slot of ``$z_7``. Simultanously ``y_8`` is no longer used and it's the top of the stack, so it's popped.
And finally

```
let x_2 := x_2_6
```
This was actually an optimization we already merged before this PR: since this is the last use of ``x_2_6``, ``x_2`` will reuse the stack slot of ``x_2_6``.

So altogether we have moved the middle argument to memory and have the other two on stack, but don't need a third slot.

Was that understandable :-)? Maybe we should extend the documentation of the mechanism a bit further in the comment in ``StackToMemoryMover.h``...
But the mechanism is to transform in chunks (e.g. one multi-variable-declaration) and, while inside the chunk we *may* need to temporarily allocate additional stack slots, after the chunk the stack is tightly packed and we have saved as many slots as we moved variables to memory...

Reply:
But yeah: the complicated part here is to be aware of the interactions of this with the stack optimization of the code transform. Only together it works for actually saving stack slots.

Reply:
Yea, maybe it's worth to make it explicit that this pretty much depends on the stack optimization

Reply:
But yea, I understand it now, thanks!